[1.x] Generate `AbstractZincFile` during `-sourcepath` workflow

Friendseeker opened the pull request:
This PR fixes #1304, #963

### Issue

https://github.com/sbt/zinc/blob/5eb0de1fdfd921f0d1b6a9302fe42593e7e27922/internal/compiler-bridge/src/main/scala/xsbt/API.scala#L62

In certain codepath, `unit.source.file` is a `PlainFile`, which causes a `MatchError`.

### Fix

When `unit.source` is built, convert `unit.source.file` to `AbstractZincFile`.

### Validating the fix

The PR includes reproduction scripted test by [lrytz](https://github.com/lrytz). The test passes after the fix.

Friendseeker commented on internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala:
```diff
@@ -78,6 +78,16 @@ sealed class ZincCompiler(settings: Settings, dreporter: DelegatingReporter, out
+    val file = f match {
+      case plainFile: PlainFile =>
+        val vf = callback.asInstanceOf[AnalysisCallback3].toVirtualFile(plainFile.file.toPath)
+        new ZincPlainFile(vf.asInstanceOf[PathBasedFile])
```
The conversion logic is... hacky.

I would appreciate any feedback suggesting a better solution.

dwijnand commented on internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala:
```diff
@@ -78,6 +79,19 @@ sealed class ZincCompiler(settings: Settings, dreporter: DelegatingReporter, out
+        virtualFile match {
+          case underlying: PathBasedFile => new ZincPlainFile(underlying)
+          case _                         => f
+        }
```
I think this is `AbstractZincFile(callback3.toVirtualFile(plainFile.file.toPath))`, btw.

dwijnand reviewed (changes requested):
See https://github.com/sbt/zinc/issues/1304#issuecomment-1875188256

Friendseeker committed:
Generate AbstractZincFile during -sourcepath workflow

In certain codepath, unit.source.file is a PlainFile, which causes a MatchError.

This PR converts PlainFile to AbstractZincFile, avoiding the MatchError.

Friendseeker commented:
@eed3si9n While the PR is in `Changes requested` status, would it be possible for you to dismiss the stale review and force merge it if everything looks okay?

My concern is that since this PR involves `AnalysisCallback3` and https://github.com/sbt/zinc/pull/1306 is already merged in, if a Zinc release is made before the PR is merged, we would need to introduce `AnalysisCallback4`.

eed3si9n commented:
Yea. Now that AnalysisCallback is AnalysisCallback3, we should just land this.
